src/keyboard.c

```diff
@@ -41,10 +41,56 @@ struct keyboard* keyboard_new(struct wl_keyboard* wl_keyboard,
 	self->wl_keyboard = wl_keyboard;
 	self->seat = seat;
 	self->context = xkb_context_new(0);
+	wl_list_init(&self->pressed_keys);
 
 	return self;
 }
 
+struct pressed_key* pressed_keys_get_key(struct wl_list* pressed_keys, uint32_t key)
+{
+	struct pressed_key* tmp;
+	wl_list_for_each(tmp, pressed_keys, link) {
+		if (key == tmp->key) {
+			return tmp;
+		};
+	}
+	return NULL;
+}
+
+void pressed_keys_add_key(struct wl_list* pressed_keys, uint32_t key)
+{
+	if (pressed_keys_get_key(pressed_keys, key)) {
+		fprintf(stderr, "Oops, already pressed key.\n");
+		return;
+	}
+	struct pressed_key* pressed_key = calloc(1, sizeof(*pressed_key));
+	assert(pressed_key);
+	pressed_key->key = key;
+	wl_list_insert(pressed_keys, &pressed_key->link);
+}
+
+void pressed_keys_remove_key(struct wl_list* pressed_keys, uint32_t key)
+{
+	struct pressed_key* tmp;
+	wl_list_for_each(tmp, pressed_keys, link) {
```

Comment:
Seems that `pressed_keys_get_key` would fit here rather naturally.

Reply:
right!